### PR TITLE
Adding language specific validation sets for Multilingual model training

### PR DESCRIPTION
**Pull request by hadyelsahar**
# Summary 
The idea of this issue to modify the [megatron-deepspeed](https://github.com/bigscience-workshop/Megatron-DeepSpeed) to track the progress of validation loss on several validaiton (periodically evaluation) sets separately. 

Currently, the validation loss is calculated on a single validation set that includes the same language combination as the training data. ([see here](https://huggingface.co/bigscience/tr1-13B-tensorboard/tensorboard) 13B param model training on tensorboard)

image

After integration of this PR user could add extra validation sets on the following form 

```
--periodic-eval-data-path \
VALID1-FR-KR 0.1 $DATA_FR 0.2 $DATA_KR, \
VALID2-JP-AR 0.2 $DATA_JP 0.3 $DATA_AR
```

Validation steps will be run automatically on each dataset independently and results will be displayed on Tensorboard
as following 

image

# What was changed
In order not to change the current way one calls the `training.py` script, I opted for adding an extra argument  `--periodic-eval-data-path`.  

Users can define extra datasets sets (EACH in a quite similar way to `--data-path` to be evaluated along with training by providing their data paths (or multiple with weights).
**Note** here that the `--split` argument does apply to the `--periodic-eval-data-path` argument. 

## Typical Examples for Multilingual training

When a model is being trained on a preprocessed multilingual dataset `multilingual`. A user can preprocess 3 monolingual datasets JP, KR, AR, and track their validation progress by sending the following argument:

```
--data-path $DATA/multilingual
--periodic-eval-data-path \
VALID-JP 1.0 $DATA_JP, \
VALID-KR 1.0 $DATA_KR, \
VALID-AR 1.0 $DATA_AR \
```

Sometimes in multilingual training, some languages are downsampled and some languages are undersampled. If a user wonders how the model performs with respect to different proportions of languages, different combinations of the languages could be passed as external validation datasets. 

```
--data-path  0.1 $DATA/EN 0.5 $DATA/JP 0.7 $DATA/KR 1.0 $DATA/AR, 
--periodic-eval-data-path \
DATASET-BALANCED 1.0 $DATA_EN 1.0 $DATA_JP 1.0 $DATA_KR 1.0 $DATA_AR, \
DATASET-NO-EN 1.0 $DATA_JP 1.0 $DATA_KR 1.0 $DATA/_R
```

# Connections with PR https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/143

https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/143 PR is developed to support the usecase:  `we can't have multilingual training data and English-only validation data at the moment.`. This use case is completely supported in this PR by adding an English-only dataset inside as a dataset to be evaluated periodically.  More over one can extend this by adding several datasets to be evaluated periodically, not just a single english only one. 

# Testing 
- Default training works  🆗 
- Integration with Tensorboard 🆗 
- Testing with real training data multiple combinations
  - 1 dataset 1 combination. 🆗 
  - 1 dataset 2 combinations 🆗 
  - 2 datasets 1 combinations 🆗 
  - 2 datasets 2 combinations 🆗 
  - 5 datasets 5 combinations 🆗 

## Independent testing @lintangsutawika  (in progress)
-

### Future Modifications (suggestions needed)
- Adding (optional) `periodic-eval-iter` and `periodic-eval-iterations` for this argument `periodic-eval-data-path`. 
if not used then fall back to the regular   `--eval-interval --eval-iters` params.

### CODE REVIEW

**Comment by sbmaruf** [comment]
Teven asked me to write this one earlier. Not sure if this solves a part of the problem discussed in this issue.

**Comment by hadyelsahar** [comment]
**Updates**

- I have added a new set of arguments `--(train|valid|test)-weighted-splits-paths` (called option2 for dataloading) that are completely orthogonal to option1 for data loading using `--data-path` and `--splits`
- fixed the logging amount of training data per language for this new data loading option2
- remove `extra-valid-data-path` that we added before
- crude testing to make sure that the previous data loading method works and the new data loading does the right thing. 

Todo: 
- [ ] deep **testing**, making sure that correct data splits are being loaded 
- [ ]  cleaner **logging** for data sizes as suggested by Stas above
 
I will be quite occupied with other stuff in the upcoming days, would be glad if somebody can take over those todos. 

Cheers all ❤️

**Comment by TevenLeScao** [comment]
I fixed a few bugs with the test iterator (ensuring `test_weighted_split_xxx`s are set to None in `data-path` mode, and a typo that passed the list of iterators instead of each iterator individually to `evaluate_and_print_results`)

**Comment by TevenLeScao** [comment]
I didn't understand why the tests I'm building were failing and now I realize that we also need to integrate this to prefix-lm. If I can do this fast I will, else we'll keep it on hold until after we finally launch an autoregressive training with this.

**Comment by TevenLeScao** [comment]
Fixed and tested.

**Comment by TevenLeScao** [comment]
So, after a really deep dive:
- The splits are respected and are non-overlapping
- The dataset weights don't seem to be always respected. It seems to me from reading the code and inspecting the objects that there are edge cases because the dataset sizes are almost always a multiple of their epoch size, so there is variation around the input weights (for example, in the extreme case, if a dataset is bigger than the amount of steps we want to train or validate on it, the entire dataset will still be passed). I have contacted the megatron team about this on Slack.

This feels like more of a Megatron issue if it is the case imo, so I vote to merge this and fix the issue independently.

@stas00

**Comment by stas00** [comment]
Based on your description it sounds that the current PR is orthogonal to the issue with datasets weights, in which case by all means if you're happy with it, please merge it. Glad you found the issue before we started training with the overlap bug.

But let's make sure that this issue you have uncovered is solved before we start training. So probably please create a new issue to track the progress.

Additionally, if you could create yet another new issue that can track all the must fix issue for 13B-ml training and add the issue from para above to it, that would be very awesome!

Thank you, Teven.

**Comment by TevenLeScao** [comment]
OK, I have confirmed with the Megatron team that this is actually not the case, the dataset weights get applied later in a C helper. Let's merge.
